hypothesis-python/src/hypothesis/extra/django/models.py

```diff
@@ -223,3 +234,20 @@ def _models_impl(draw, strat):
         return draw(strat)[0]
     except IntegrityError:
         reject()
+
+
+@st.composite
+def _models_with_pk_impl(draw, model, primary_key, kwargs):
+    """Draw a value for a model where the primary key is drawn as part of the
+    strategy."""
+    constructor_args = {}
```

Comment:
I'd like to move the dictionary handling out of this impl function - it's mostly a style issue, but also that the reason to have these separately is to minimize the work redone to draw data for each test case (and thus performance overhead).  Above, you could write a shorter version:

```python
kwargs = {field: result.pop(field)}
kwargs['defaults'] = result
return _models_with_pk_impl(model, field, kwargs)
```

Reply:
I'm not sure I'm following what you're suggesting. We have to create a fresh dictionary each time `_models_with_pk_impl` is called, as far as I can see: the values in the dictionary `constructor_args['defaults']` will be different each time. If we pass in the dictionary as you describe (dict of dicts of strategies), we'll still need to iterate over `kwargs['defaults']` and create a new dict where each value is the result of drawing from the strategies in the source dict. I don't see an efficiency gain.

Style-wise, it feels like the work of building up the arguments to the `update_or_create` and dealing with its idiosyncrasies should be done nearest the call to `update_or_create`, all other things being equal.

The thing I'm not totally comfortable with (which may actually be what you're driving at) is that I'm trying to incorporate the work normally done by `strategies.builds` into this function. Separating the two layers out (drawing from a dict-of-dicts-of-strategy, and wrangling the Django code) feel like they could be separated conceptually, but I don't see much potential for reuse. I'm happy to have a go at refactoring if you think this is a better way to go.

Reply:
Gah.  I'd typed a much longer response, then lost it.  The short version is that I'd prefer a declarative format, and (I think) we can actually keep using the current impl function as our only way of handling database errors.  Please correct me if I'm wrong!

```python
kwargs = {field: result.pop(field)}
kwargs['defaults'] = st.fixed_dictionaries(result)
return _models_impl(st.builds(model.objects.update_or_create, **kwargs))
```